Test that DX coil performance curves are near 1.0 at rated condition and VRF coils use terminal unit flow rate for sizing

rraustad opened the pull request:
Pull request overview
---------------------
Input files with bad performance curves for DX coils can cause odd results (e.g., very cold zone temperatures). Add test of performance curves and report warning when curve result is not near 1.0 at reference condition. Additionally, the VRF TU allows input for air flow rate, this value should also be used to size the VRF DX coils. This corrects issue #6118.

Note: now that performance curves are tested, some existing files may warn that performance curves are not normalized to 1 at the rating condition. This is not a cause for concern if the performance curves were normalized to an off-design rating point.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Code style (parentheses padding, variable names)
 - [x] Functional code review (it has to work!)
 - [x] If defect, results of running current develop vs this branch should exhibit the fix
 - [x] CI status: all green or justified
 - [x] Performance: CI Linux results include performance check -- verify this
 - [x] Unit Test(s)
 - C++ checks:
   - [x] Argument types
   - [x] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes: n/a
 - [x] Documentation changes in place
 - [x] Changed docs build successfully

rraustad committed:
Add tests to ensure user performance curves are valid for DX coils

rraustad commented:
New warning message when performance curves are out of bounds:

    ** Warning ** GetDXCoils: Coil:Cooling:DX:VariableRefrigerantFlow="THERMAL ZONE 1-3|VTU-1|VRF COOLING COIL"
    **   ~~~   ** ...Cooling Capacity Ratio Modifier Function of Temperature Curve Name output is not equal to 1.0 (+ or - 10%) at reference conditions.
    **   ~~~   ** ...Curve output at reference conditions = 23.000

This CapFT curve multiplied the DX coil capacity 23 fold and resulted in outlet air temps of -272 C.

Existing warning message attempted to direct the user to this coil but it really did not help diagnose the problem.

    ** Warning ** Coil:Cooling:DX:VariableRefrigerantFlow "THERMAL ZONE 1-3|VTU-1|VRF COOLING COIL" - Full load outlet air dry-bulb temperature < 2C. This indicates the possibility of coil frost/freeze. Outlet temperature = -17.77 C.
    **   ~~~   **  ...Occurrence info = SAN DIEGO MONTGOMER ANN CLG 1% CONDNS DB=>MWB, 08/21 07:00 - 07:01
    **   ~~~   ** ... Possible reasons for low outlet air dry-bulb temperatures are: This DX coil
    **   ~~~   **    1) may have a low inlet air dry-bulb temperature. Inlet air temperature = 22.830 C.
    **   ~~~   **    2) may have a low air flow rate per watt of cooling capacity. Check inputs.
    **   ~~~   **    3) is used as part of a HX assisted cooling coil which uses a high sensible effectiveness. Check inputs.

Not sure what else if anything can be done for this type of input error.

rraustad committed:
Allow zone equipment parent to size child cooling and heating air flow rates.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #6118-VRF-Sizing-Problem-and-PLR-error

rraustad commented:
Now there are lots of these. Either we never knew we were using a bad curve, or the curve is good and something else is amiss.

    +   ** Warning ** GetDXCoils: Coil:Heating:DX:SingleSpeed="HEAT PUMP DX HEATING COIL 1"
    +   **   ~~~   ** ...Energy Input Ratio Function of Temperature Curve Name output is not equal to 1.0 (+ or - 10%) at reference conditions.
    +   **   ~~~   ** ...Curve output at reference conditions = 0.819

rraustad commented:
Something was amiss. It's not the curves it's the programming. Cubic and Quadratic are only a function of OAT. Copy-paste and forgot to remove Indoor T and the curve manager used that value.

    EIR = DXCoil( DXCoilNum ).RatedEIR( Mode ) * EIRTempModFac * EIRFlowModFac;

    if ( ( DXCoil( DXCoilNum ).EIRTempModFacCurveType( 1 ) == Quadratic ) || ( DXCoil( DXCoilNum ).EIRTempModFacCurveType( 1 ) == Cubic ) ) {
        EIRTempModFac = CurveValue( DXCoil( DXCoilNum ).EIRFTemp( Mode ), OutdoorDryBulb );
    } else if ( DXCoil( DXCoilNum ).EIRTempModFacCurveType( 1 ) == BiQuadratic ) {
        EIRTempModFac = CurveValue( DXCoil( DXCoilNum ).EIRFTemp( Mode ), InletAirDryBulbTemp, OutdoorDryBulb );
    } else {


    if ( SELECT_CASE_var == "QUADRATIC" ) {
        CurveVal = CurveValue( DXCoil( DXCoilNum ).CCapFTemp( 1 )), RatedInletAirTempHeat, RatedOutdoorAirTempHeat );
    } else if ( SELECT_CASE_var == "BIQUADRATIC" ) {
        CurveVal = CurveValue( DXCoil( DXCoilNum ).CCapFTemp( 1 ), RatedInletAirTempHeat, RatedOutdoorAirTempHeat );
    } else if ( SELECT_CASE_var == "CUBIC" ) {
        CurveVal = CurveValue( DXCoil( DXCoilNum ).CCapFTemp( 1 )), RatedInletAirTempHeat, RatedOutdoorAirTempHeat );

![dxheatperfcurves](https://cloud.githubusercontent.com/assets/5778156/25961884/c9aa86ec-3649-11e7-8cd5-06da6a8e9213.png)

rraustad committed:
Correct programming errors in performance curve testing in DX coils

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -2283,10 +2287,10 @@ namespace DXCoils {
 
 				if ( SELECT_CASE_var == "QUADRATIC" ) {
-					// curve limits at rating point tested below
+					// curve ouput tested below
```
sowy about my lisp, I ouput incowectly

rraustad commented:
I guess it just figures that fixing this VRF problem would highlight issues with widely used performance curves ONLY IN VRF input files!

rraustad commented:
Well, I see where the VRF TU CapFT curves came from and the curve coefficients match the spreadsheet. What I don't understand is why Mitsubishi used 16C as the reference point (Figure 1 [here](http://www.fsec.ucf.edu/en/Publications/pdf/FSEC-CR-1910-12.pdf)). The Daikin table data in Figure 2 shows 19C as the reference point with 33.5 kW cooling capacity reference (grey highlight in Fig 2). To fix this I would need to go back and adjust the performance curves to use the reference capacity at 19.4C IWB.

![vrftucapft](https://cloud.githubusercontent.com/assets/5778156/26004221/0cf4eafa-3703-11e7-90b2-fd507c045fda.png)

rraustad committed:
Add check for VRF condenser CapFT

mjwitte commented:
@rraustad So, what's the plan here.  Are we going to leave these curves as-is, knowing that the rated capacity is off a bit?  Is there not a standard rating condition for VRF units?

mjwitte commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -3942,7 +4142,7 @@ namespace DXCoils {
 						if ( CurveVal > 1.10 || CurveVal < 0.90 ) {
 							ShowWarningError( RoutineName + CurrentModuleObject + "=\"" + DXCoil( DXCoilNum ).Name + "\", curve values" );
-							ShowContinueError( "..." + cAlphaFields( 13 + ( I - 1 ) * 6 ) + " output is not equal to 1.0 (+ or - 10%) at rated conditions." );
+							ShowContinueError( "..." + cAlphaFields( 13 + ( I - 1 ) * 6 ) + " output is not equal to 1.0 (+ or - 10%) at reference conditions." );
```
@rraustad Why is this one reference instead of rated?  It's using fixed Rated temps in the curve call.

rraustad replied:
Existing checks of curves used "reference" in the message. I used "rated". I searched and replaced all with "rated".

mjwitte commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -3972,6 +4172,7 @@ namespace DXCoils {
 						CurveVal = CurveValue( DXCoil( DXCoilNum ).MSCCapFFlow( I ), 1.0 );
 						if ( CurveVal > 1.10 || CurveVal < 0.90 ) {
 							ShowWarningError( RoutineName + CurrentModuleObject + "=\"" + DXCoil( DXCoilNum ).Name + "\", curve values" );
+							ShowContinueError( cAlphaFields( 14 + ( I - 1 ) * 6 ) + " = " + Alphas( 14 + ( I - 1 ) * 6 ) );
```
Field name is redundant here and in the next line.  This line is nice, but then I'd remove the field name in the next line.

There's so many of these, all using about six lines each.  Seems ripe for a separate function.
```
CheckCuveAtRatedConditions( RoutineName + CurrentModuleObject, DXCoil( DXCoilNum ).Name, DXCoil( DXCoilNum ).MSCCapFFFlow, cAlphaFields( . . . ), X1, X2-optional, X3-optional );
```

mjwitte commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -12836,6 +13108,71 @@ Label50: ;
+		// DX Coil is not a legal DX Coil.
+
+		// METHODOLOGY EMPLOYED:
+		// na
```
Remove this and other unused comment blocks below.

mjwitte commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -5408,7 +5529,10 @@ namespace HVACVariableRefrigerantFlow {
+			ZoneEqSizing( CurZoneEqNum ).CoolingAirFlow = true;
+			ZoneEqSizing( CurZoneEqNum ).CoolingAirVolFlow = VRFTU( VRFTUNum ).MaxCoolAirVolFlow;
+			ZoneEqSizing( CurZoneEqNum ).HeatingAirFlow = true;
+			ZoneEqSizing( CurZoneEqNum ).HeatingAirVolFlow = VRFTU( VRFTUNum ).MaxHeatAirVolFlow;
```
Looks like there are more changes here than just adding a bunch of curve value checks?  Please describe this and any other changes in the PR synopsis.

mjwitte replied:
And change the title.

rraustad replied:
Changes in response to: MJWitte [wrote](https://github.com/NREL/EnergyPlus/issues/6118) - Ok, but what about item 1 above, the coil sizing when there's a hard flow rate in the terminal unit?

rraustad commented:
Added new function with output as:

    ** Warning ** GetDXCoils: Coil:Cooling:DX:VariableRefrigerantFlow="TU1 VRF DX COOLING COIL curve values
    **   ~~~   ** ... Cooling Capacity Ratio Modifier Function of Temperature Curve Name = VRFTUCOOLCAPFT output is not equal to 1.0 (+ or - 10%) at rated conditions.
    **   ~~~   ** ... Curve output at rated conditions = 1.136

Old warning shows:

    ** Warning ** GetDXCoils: Coil:Cooling:DX:VariableRefrigerantFlow="TU1 VRF DX COOLING COIL"
    **   ~~~   ** ...Cooling Capacity Ratio Modifier Function of Temperature Curve Name output is not equal to 1.0 (+ or - 10%) at rated conditions.
    **   ~~~   ** ...Curve output at rated conditions = 1.136

rraustad commented:
@Myoldmopar what do I need to do to get rid of optional "_" in my call? CurveValue call doesn't need it and can take up to 4 additional arguments? Maybe you need to wait to see my new function?

    if ( SELECT_CASE_var == "CUBIC" ) {
        checkCurveValue( RoutineName + CurrentModuleObject, DXCoil( DXCoilNum ).Name, DXCoil( DXCoilNum ).HCOPFAirFlow, cAlphaFields( 14 ), Alphas( 14 ), 1.0, _, _, _, _ );
    } else if ( SELECT_CASE_var == "QUADRATIC" ) {
        HeatCOPFAirFlow = CurveValue( DXCoil( DXCoilNum ).HCOPFAirFlow, 1.0 );

rraustad committed:
Added new function to test performance curves, and removed unnecessary comment lines in DXCoils

rraustad commented:
I addressed all comments, thank you Mike.

Re the new errors in VRF example files, those are valid curves just set up at off-design conditions. So I am not sure what to do about that. Add a disclaimer somewhere about off-design curves, re-engineer the example file curves?
That VRF system was deployed before AHRI 1230, but manufacturers were supposed to be using AHRI 210/240 at that time, so I am still confused why the rating point is at 16C and not 19.4C. I could fix the example file but that would take some time.

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -3177,30 +3068,17 @@ namespace DXCoils {
 					{ auto const SELECT_CASE_var( GetCurveType( DXCoil( DXCoilNum ).HCOPFTemp ) );
 
-					if ( SELECT_CASE_var == "BIQUADRATIC" ) {
+					if ( SELECT_CASE_var == "BIQUADRATIC" || SELECT_CASE_var == "CUBIC" ) {
```
oops! curve type is needed here

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -12785,53 +12237,77 @@ Label50: ;
 			GetCoilsInputFlag = false;
 		}
 
-		DXCoilIndex = FindItemInList( DXCoilName, DXCoil );
```
Git just confused here, this line wasn't deleted

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -12785,53 +12237,77 @@ Label50: ;
+					ShowSevereError( ThisObjectType + ", GetDXCoilIndex: DX Coil not found " );
 				} else {
-					ShowSevereError( "GetDXCoilIndex: DX Coil not found=" + DXCoilName );
+					ShowSevereError( "GetDXCoilIndex: DX Coil not found " );
```
same here, similar lines in function GetDXCoilIndex and GetDXCoilName are getting munged.

mjwitte commented on src/EnergyPlus/CurveManager.hh:
```diff
@@ -571,6 +571,20 @@ namespace CurveManager {
+		std::string const cFieldName,     // object field name
+		std::string const cFieldValue,    // user input curve name
+		Real64 const Var1,                // required 1st independent variable
+		Optional <Real64 const > Var2,    // 2nd independent variable
```
@rraustad I think you can avoid the need for trailing "_" in the calls by providing a default here
`Optional <Real64 const > Var2 = "_", // 2nd independent variable`
Like [CurveManager.hh](https://github.com/NREL/EnergyPlus/blob/173bb58bcc95ec005c91946ecee8f573224904fd/src/EnergyPlus/CurveManager.hh#L412-L418)

rraustad committed:
Correct inadvertent deletion of code. Update documentation. Use new curve value function in VRF parent. Delete optional fields in call. Update example files with note about off-design rating point.

rraustad committed:
Minor change to IO Ref

rraustad committed:
Minor update to IO Ref

mjwitte commented on src/EnergyPlus/CurveManager.cc:
```diff
@@ -6379,6 +6379,42 @@ Label999: ;
+		std::string const objectName,        // parent object where curve is used
+		int const curveIndex,                // index to curve object
+		std::string const cFieldName,        // object field name
+		std::string const cFieldValue,       // user input curve name
```
This argument isn't really needed, since the curve name is known from the index, but probably not worth changing all of these now.  Otherwise, much better with this new function.

rraustad replied:
Good point, it's only 68 places so if I do need another change, this one should be included.

mjwitte replied:
**Only** 68?

Myoldmopar commented:
> @Myoldmopar what do I need to do to get rid of optional "_" in my call? CurveValue call doesn't need it and can take up to 4 additional arguments? Maybe you need to wait to see my new function?

Maybe it's just getting late, but I think you should be able to just not pass them in if they have default values in the function signature.
